compiler/inference/dataflow.py

```diff
@@ -0,0 +1,107 @@
+from collections.abc import Generator, Iterable, Sequence
+from dataclasses import dataclass
+
+from xdsl.dialects import scf
+from xdsl.ir import Block, Operation, SSAValue, Use
+from xdsl.rewriter import InsertPoint
+from xdsl.traits import IsTerminator
+
+
+@dataclass(frozen=True, slots=True)
+class InsertCandidate:
+    idx: int
+    op: Operation
+
+
+def uses_through_controlflow(val: SSAValue) -> Generator[Use, None, None]:
+    """
+    Find all uses of an SSA value, even through control flow. For example, given the IR:
+
+    ```
+    %a = arith.constant 1 : i32
+
+    %res = scf.for (...) iter_args(%x = %a : i32) {
+        %loop = arith.add ..., %x
+        yield %x
+    }
+
+    arith.add %a, ...
+    ```
+
+    uses_through_controlflow(%a) will yield:
+        - scf.for
+        - %loop = arith.add ..., %x
+        - yield %x
+        - arith.add %a, ...
+    """
+    for use in val.uses:
+        # return each use
+        yield use
+        # if it's an scf.for, recurse on block arg
+        if isinstance(use.operation, scf.For):
+            for_op = use.operation
+            yield from uses_through_controlflow(for_op.body.block.args[use.index - 2])
+        # if it's a yield, recurse on parent op result
+        elif isinstance(use.operation, scf.Yield):
+            yield from uses_through_controlflow(
+                use.operation.parent_op().results[use.index]
+            )
+
+
+def get_insertion_points_where_val_dangles(
+    vals: Sequence[SSAValue],
+) -> Iterable[InsertPoint]:
+    """
+    Return all insertion points for positions where a value starts dangling.
+
+    If val is used by a terminator, it's not assumed dangling in that block.
+
+    If multiple vals are given, return the earliest point in each block where
+    all vals used in that block are dangling.
```

Comment:
Why would we need to have multiple vals here? Is this because we both wait for the token and the state?
~~I find this quite difficult to read, would it make more sense to not make this a generator instead?~~
I should just read more generators haha